File: 0019-remove-nt-node-from-end-of-list/main.go
package main
import ("fmt")

type ListNode struct {
  Val int
  Next *ListNode
}

// beats 100%
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	// fmt.Println("n: ", n)
	_head := head

	flatListLen := 0
	for head != nil {
		flatListLen += 1
		head = head.Next
	}
	flatList := make([]*ListNode, flatListLen)

	// fmt.Println("flatList: ", flatList)

	head = _head // reset

	i := 0
	for head != nil {
		flatList[i] = head
		head = head.Next
		i++
	}

	fmt.Println("flatList: ", flatList)

	flatList2 := make([]*ListNode, flatListLen)

	if flatListLen - n == 0 {
		// copy all except first
		for i := 1; i < len(flatList); i++ {
			flatList2[i - 1] = flatList[i]
		}
	} else {
		// copy all
		for i := 0; i < flatListLen; i++ {
			flatList2[i] = flatList[i]
		}
		a := flatList2[flatListLen - n - 1]
		b := flatList2[flatListLen - n]
		a.Next = b.Next
	}

	return flatList2[0]
}

func main() {
	e1 := &ListNode{Val: 1, Next: &ListNode{Val: 2, Next: &ListNode{Val: 3, Next: &ListNode{Val: 4, Next: &ListNode{Val: 5}}}}}

	r := removeNthFromEnd(e1, 2)

	fmt.Println(r.Val)
	fmt.Println(r.Next.Val)
	fmt.Println(r.Next.Next.Val)
	fmt.Println(r.Next.Next.Next.Val)


	e2 := &ListNode{Val: 1}
	r = removeNthFromEnd(e2, 1)
	fmt.Println(r == nil)
}

File: 0019-remove-nt-node-from-end-of-list/main.js
function ListNode(val, next) {
  this.val = (val===undefined ? 0 : val)
  this.next = (next===undefined ? null : next)
}

// beats 100%
var removeNthFromEnd = function(head, n) {
  let flatList = [];
  while (head) {
    flatList.push(head);
    head = head.next;
  }

  console.log("flatList", flatList)

  if (flatList.length - n === 0) {
    console.log("flatList.length - n === 0")

    flatList.shift()
  } else {
    let a = flatList[flatList.length - n - 1];
    console.log("a", a);

    let b = flatList[flatList.length - n];
    console.log("b", b);

    a.next = b.next;
  }

  console.log("flatList", flatList)

  return flatList[0] || null;
};

// let a = [1,2,3,4,5];
// let b = 2;
// let n = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));

let a = [1];
let b = 1;
let n = new ListNode(1);

let r = removeNthFromEnd(n, b);
console.log("r", r);